### ISSUE TITLE
searchableAttributes doesn't support full email addresses

### ISSUE THREAD

**Issue by BenParizek** [craftcms/cms#3464]
### Description

I have a custom Element that has defined an email column as a searchable attribute. The column gets indexed and can be searched but only if the "." is not included in the email address used in the search. 

``` php
protected static function defineSearchableAttributes(): array
{
        return ['emailAddressColumn'];
}
```

Testing this with the default Users Element the behavior seems to be the same for the email column.

No search results appear:

- person@domain.com
- domain.com 

Results appear:

- person
- person@
- person@domain
- domain
- .com (this is the only search that I've found where the dot works)

### Additional info

- Craft version: 3.0.30.2
- PHP version: 7.2
- Database driver & version: MySQL
- Plugins & versions: several under development

**Comment by brandonkelly** [comment]
Are you using real world emails or literally _person@domain.com_ ? If the latter, it could be due to MySQL fulltext stop words.

**Comment by BenParizek** [comment]
I'm using real world emails, I just updated the emails for the examples posted here. I tested both our business email address as well as a gmail address and seemed to get the same results for both.

**Comment by angrybrad** [comment]
Turns out searching is difficult.

Before Craft saves search index keywords into the `searchindex` table, it will run each of those though a normalization process which includes removing punctuation including several other checks:

https://github.com/craftcms/cms/blob/develop/src/helpers/Search.php#L41-L81

It also runs search query keywords through the same normalization request when comparing against the `searchindex` table.

For example, this `person@domain.com` will be inserted into the search index table as ` person domain com ` and when you search with `person@domain`, Craft is actually using `person domain` to make the comparison.

When you get results using `person` and `person@`, that's effectively the same search as the `@` will get stripped out and match ` person ` in the index using a full-text search.

Getting a result using `person@domain` (`person domain`) matches ` person domain ` in the search index.

Getting a result using `domain` matches ` domain` in the search index.

Getting a result using `.com` (which gets normalized to `com`) is different.  MySQL by default is set to ignore searches in a full text search using MyISAM that are under a certain number of characters (4 by default https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_ft_min_word_len)

Craft sees this and changes from using a full-text search to MySQL `LIKE`, `NOT LIKE` in there `WHERE` part of the clause as a workaround. Less flexible and efficient, but O.K. for simple cases, so that's how you get the match.

Why these don't match:

    person@domain.com
    domain.com

My guess is that it's because Craft decides to use the full text search because of overall length, but the last part of the segment `com` falls below the MySQL minimum full text word length, so it doesn't match any result.

If you adjusted `ft_min_word_len` on the server to a lower value, they'd probably show up, but you'll possibly get extra noise in the results as well.

Alternatively, if you're running at least MySQL 5.6, InnoDB added support for full-text searching that doesn't have a minimum word length as far as I can tell.  I swapped the `searchindex` table over to that and both `person@domain.com` and `domain.com` returned a match.

https://dev.mysql.com/doc/refman/5.6/en/innodb-fulltext-index.html

We can't use that by default in Craft 3, because we support back to MySQL 5.5.

Ultimately we'll be able to get around weirdness like this by implementing https://github.com/craftcms/cms/issues/833
